### DISCUSSION TITLE
Notify client when property value status code has been changed

### DISCUSSION THREAD

**Discussion by che2116** [FreeOpcUa/opcua-asyncio#401] [Q&A] [answered]
Is there a way on the server side to notify a client when the property value status code has been changed, but the property value not?

**Comment by AndreasHeine** [comment]
in general only if the client has subscribed it!

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-subscription.py

**Reply by che2116** [reply]
I understand, that  the client has to subsribe to the data change event. But my question is about the server side. To be more precise, here is some code from my server:
```
async def async_main():
    # setup server
    server = Server()
    await server.init()
    server.set_endpoint('opc.tcp://0.0.0.0:4840/freeopcua/server/')
    uri = 'http://examples.freeopcua.github.io'
    idx = await server.register_namespace(uri)

    #Create or import from xml object types
    dev = await server.nodes.base_object_type.add_object_type(idx, "MyDevice")
    await (await dev.add_property(idx, "sensor1", 0, ua.VariantType.UInt16)).set_modelling_rule(True)
    #Create objects in the OPC UA address space
    mydevice = await server.nodes.objects.add_object(idx, "Device0001", dev)
    sensor1_val = mydevice.get_child(f"{idx}:sensor1")
    # Here is questions: 
    # - How can i know that the client has subscribed to the sensor1_val data chganged event?
    # - What is the sampling interval? 

    _logger.info('Starting server!')
    async with server:
        while True:
            # poll real device and update sensor1_val value
            # But if i update it value with sensor1_val.write_value()
            # and pass ua.DataValue as argument 
            # client will be notified only if data_value.Value has been changed
            # if change only data_value.StatusCode nothing will happen
```

**Reply by AndreasHeine** [reply]
> How can i know that the client has subscribed to the sensor1_val data chganged event?

actually no easy way to get it! if you want to find it out you need to digg deep into the server class... there must be a better way! why is it important? i would use a asyncio.task/coroutine to update vars

> What is the sampling interval?

from spec. side -> https://reference.opcfoundation.org/v104/Core/docs/Part4/5.12.1/#5.12.1.2

> if change only data_value.StatusCode nothing will happen

if the DataValue Class instance is the same then Yes if you want to trigger a notification you need to make a new instance with the old value and another status code

this guy wants it the other way around but the principle is the same:
https://github.com/FreeOpcUa/python-opcua/discussions/1195#discussioncomment-185866

**Comment by che2116** [comment]
> if the DataValue Class instance is the same then Yes if you want to trigger a notification you need to make a new instance with the old value and another status code

That is not working. Here is short example:
```python
import asyncio
import copy
import logging
from datetime import datetime
from asyncua import ua, Server

async def update_node_val(ua_node, value, status):
    #The new instance of the ua.DataValue class creates on every call
    data_val = ua.DataValue(ua.Variant(value, ua.VariantType.Int16),\
        status=ua.StatusCode(status), sourceTimestamp=datetime.now(),serverTimestamp=datetime.now())
    await ua_node.write_value(data_val)

async def main():
    server = Server()
    await server.init()
    server.set_endpoint("opc.tcp://0.0.0.0:4840/freeopcua/server/")
    server.set_server_name("FreeOpcUa Example Server") 
    uri = "http://examples.freeopcua.github.io"
    idx = await server.register_namespace(uri)
    dev = await server.nodes.base_object_type.add_object_type(idx, "MyDevice")
    await (await dev.add_property(idx, "device_id", 340, ua.VariantType.Int16)).set_modelling_rule(True)
    mydevice = await server.nodes.objects.add_object(idx, "Device0001", dev)
    dev_id = await mydevice.get_child(f"{idx}:device_id")
    async with server:
        i = 0
        while True:
            await asyncio.sleep(1)
            if i % 10 < 5:
                await update_node_val(dev_id, 100, ua.StatusCodes.Good)
            else:
                #Nothing will happen on the client side
                await update_node_val(dev_id, 100, ua.StatusCodes.BadNotConnected)
                # This will trigger a notification
                # await update_node_val(dev_id, 101, ua.StatusCodes.BadNotConnected)
            i += 1

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(main())
```

> why is it important? i would use a asyncio.task/coroutine to update vars

Yes, I do it in the same way. But I think it would be nice if I could know exactly what data the client needs and how often. So I would be able to poll only those devices for which data is needed.

**Reply by AndreasHeine** [reply]
ok! the problem with BadStatusCodes is the Value is not usable/valid (and should not be used for anything, i am still looking there was some best practice for BadStatusCodes in the OPC UA Spec. stilll looking...hang on) so you dont really know the true value in this case you can just write 0 -> Value 0 and BadNotConnected and you know there is an issue (on clientside you can store the last known Good Value and there you go!?)
another way would be to use something like an custom structure for your devicetype which containes 2 filds (value, devicestatus)

**Comment by che2116** [comment]
I don't understand why the server should not notify the client when the UA Node changed it's status from good to bad and vice versa? And if the server shoud notify the client about the node status changing, than from what status to what?

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/v104/Core/docs/Part4/7.34.1/
All OPC UA Clients shall always check the StatusCode associated with a result before using it. Results that have an uncertain/warning status associated with them shall be used with care since these results might not be valid in all situations. **Results with a bad/failed status shall never be used**.

https://reference.opcfoundation.org/v104/Core/docs/Part8/6.3.2/
Bad_NotConnected: The variable should receive its value from some data source, but has never been configured to do so.

**Comment by che2116** [comment, marked as answer]
First of all thank you  for  answers. Next, summarizing all your answers we have:
The server must not notify the client about changing a monitoring item status or timestamps (source timestamp or server timestamp), when it's value has not been changed. This is part of the OPC UA specification.
